src/mo_frontend/typing.ml

```diff
@@ -2204,31 +2209,153 @@ and infer_call env exp1 inst exp2 at t_expect_opt =
+        Idea:
+        - Decompose the argument into sub-expressions and defer inference for those that would fail.
+        - Find a partial instantiation first using the inferred sub-expressions.
+        - Substitute and proceed with the remaining sub-expressions to get the full instantiation.
+       *)
+      let infer_subargs_for_bimatch_or_defer env exp target_type =
+        let rec cannot_infer_pat pat =
+          match pat.it with
+          | WildP
+          | VarP _
+          | AltP _ -> true (* cases that cannot be inferred, would report errors *)
+          | LitP _
+          | AnnotP _ (* annotated patterns can be inferred *)
+          | SignP _ -> false
+          | TupP pats -> List.exists cannot_infer_pat pats
+          | ParP p
+          | OptP p
+          | TagP (_, p) -> cannot_infer_pat p
+          | ObjP pfs -> List.exists cannot_infer_pat_field pfs
+        and cannot_infer_pat_field pf =
+          match pf.it with
+          | ValPF (_, pat) -> cannot_infer_pat pat
+          | TypPF _ -> false
+        in
+        let try_infer_exp env exp =
+          match exp.it with
+          | FuncE (_, _, _, pat, _, _, _) when cannot_infer_pat pat -> None
+          (* Future work: more cases *)
+          | _ -> Some (infer_exp env exp)
+        in
+        let infer_or_defer env (subs, deferred, to_fix) exp target_type =
+          match try_infer_exp env exp with
+          | Some t -> t, ((t, target_type) :: subs, deferred, to_fix) (* subtype problem for bi_match *)
+          | None -> target_type, (subs, (exp, target_type) :: deferred, to_fix) (* deferred *)
+        in
+        let rec decompose env exp target_type acc =
+          match exp.it, T.unwrap_named target_type with
+          | TupE exps, T.Tup ts when List.length exps = List.length ts ->
+            let ts', (subs, deferred, to_fix) = decompose_list env exps ts [] acc in
+            let target_type' = T.Tup ts' in
+            (* exp.note would need to be fixed later after the substitution *)
+            target_type', (subs, deferred, (exp, target_type') :: to_fix)
+          (* Future work: more cases *)
+          | _ -> infer_or_defer env acc exp target_type
+        and decompose_list env exps ts acc_ts acc =
+          match exps, ts with
+          | exp::exps, t::ts ->
+            let t', acc' = decompose env exp t acc in
+            decompose_list env exps ts (t' :: acc_ts) acc'
+          | [], [] -> List.rev acc_ts, acc
+          | _ -> assert false
+        in
+        decompose env exp target_type ([], [], [])
+      in
+
+      (* Infer the argument as much as possible, defer sub-expressions that cannot be inferred *)
+      let (t2, (subs, deferred, to_fix)) = infer_subargs_for_bimatch_or_defer env exp2 t_arg in
+      if Bi_match.debug then begin
+        print_endline (Printf.sprintf "exp2 : %s" (Source.read_region_with_markers exp2.at |> Option.value ~default:""));
```

Comment:
The debug spew is probably helpful, but can you abstract the code into functions to make the real logic easier to see?